File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        
        function dfs(node){
            if(!node) return false;

            if(sameTree(node, subRoot)) return true;
            return dfs(node.left) || dfs(node.right);
        }

        function sameTree(n, s){
            if(!n && !s) return true;
            if(!n || !s) return false;
            if(n.val != s.val) return false;
            return sameTree(n.left, s.left) && sameTree(n.right, s.right);
        }

        if(!subRoot) return false;
        return dfs(root);
    }
}

File: Data Structures & Algorithms/house-robber-iii/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        dp = {}

        def dfs(node):
            if node in dp:
                return dp[node]
            if not node:
                return 0, 0

            # using this, not

            ret_l, ret_r = 0, 0

            l_l, l_r = dfs(node.left)
            r_l, r_r = dfs(node.right)
            
            ret_l = node.val + l_r + r_r
            ret_r = max(l_l, l_r) + max(r_l, r_r)

            dp[node] = ret_l, ret_r
            return dp[node]

        return max(dfs(root))


